subsys/logging/log_core.c

```diff
@@ -28,6 +28,161 @@ static bool initialized;
+		 | ((((message >> 7)  & 0x01) << (pin_nums->BIT7))));
+	}
+	return mask;
+}
+
+uint32_t create_mask_gpio_of_pins(struct LOG_FRONTEND_PIN_NUMBERS * pin_nums)
+{
+	uint32_t mask_of_pins = create_mask_gpio(pin_nums, 255);
+	mask_of_pins = mask_of_pins | (0x00000001 << (pin_nums->CLK));
+	return mask_of_pins;
+}
+
+void send_message_gpio_8b(uint8_t message, struct LOG_FRONTEND_PIN_NUMBERS * pin_nums, uint32_t mask_of_pins)
+{
+	uint32_t mask_to_set = create_mask_gpio(pin_nums, message);
+	nrf_gpio_port_out_set(LOG_FRONTEND_GPIO_REG_PTR, mask_to_set);
+	nrf_gpio_port_out_clear(LOG_FRONTEND_GPIO_REG_PTR, (mask_to_set ^ mask_of_pins));
+	nrf_gpio_pin_set(NRF_GPIO_PIN_MAP(CONFIG_LOG_FRONTEND_PORT_NUMBER, pin_nums->CLK));
+}
+
+void send_message_gpio_16b(uint16_t message, struct LOG_FRONTEND_PIN_NUMBERS * pin_nums, uint32_t mask_of_pins)
+{
+	uint32_t mask_to_set = create_mask_gpio(pin_nums, message & 0xff);
+	nrf_gpio_port_out_set(LOG_FRONTEND_GPIO_REG_PTR, mask_to_set);
+	nrf_gpio_port_out_clear(LOG_FRONTEND_GPIO_REG_PTR, (mask_to_set ^ mask_of_pins));
+	nrf_gpio_pin_set(NRF_GPIO_PIN_MAP(CONFIG_LOG_FRONTEND_PORT_NUMBER, pin_nums->CLK));
+
+	mask_to_set = create_mask_gpio(pin_nums, (message >> 8) & 0xff);
+	nrf_gpio_port_out_set(LOG_FRONTEND_GPIO_REG_PTR, mask_to_set);
+	nrf_gpio_port_out_clear(LOG_FRONTEND_GPIO_REG_PTR, (mask_to_set ^ mask_of_pins));
+	nrf_gpio_pin_set(NRF_GPIO_PIN_MAP(CONFIG_LOG_FRONTEND_PORT_NUMBER, pin_nums->CLK));
+}
+
+void send_message_gpio_32b(uint32_t message, struct LOG_FRONTEND_PIN_NUMBERS * pin_nums, uint32_t mask_of_pins)
+{
+	uint32_t mask_to_set = create_mask_gpio(pin_nums, message & 0xff);
+	nrf_gpio_port_out_set(LOG_FRONTEND_GPIO_REG_PTR, mask_to_set);
+	nrf_gpio_port_out_clear(LOG_FRONTEND_GPIO_REG_PTR, (mask_to_set ^ mask_of_pins));
+	nrf_gpio_pin_set(NRF_GPIO_PIN_MAP(CONFIG_LOG_FRONTEND_PORT_NUMBER, pin_nums->CLK));
+
+	mask_to_set = create_mask_gpio(pin_nums, (message >> 8) & 0xff);
+	nrf_gpio_port_out_set(LOG_FRONTEND_GPIO_REG_PTR, mask_to_set);
+	nrf_gpio_port_out_clear(LOG_FRONTEND_GPIO_REG_PTR, (mask_to_set ^ mask_of_pins));
+	nrf_gpio_pin_set(NRF_GPIO_PIN_MAP(CONFIG_LOG_FRONTEND_PORT_NUMBER, pin_nums->CLK));
+
+	mask_to_set = create_mask_gpio(pin_nums, (message >> 16) & 0xff);
+	nrf_gpio_port_out_set(LOG_FRONTEND_GPIO_REG_PTR, mask_to_set);
+	nrf_gpio_port_out_clear(LOG_FRONTEND_GPIO_REG_PTR, (mask_to_set ^ mask_of_pins));
+	nrf_gpio_pin_set(NRF_GPIO_PIN_MAP(CONFIG_LOG_FRONTEND_PORT_NUMBER, pin_nums->CLK));
+
+	mask_to_set = create_mask_gpio(pin_nums, (message >> 24) & 0xff);
+	nrf_gpio_port_out_set(LOG_FRONTEND_GPIO_REG_PTR, mask_to_set);
+	nrf_gpio_port_out_clear(LOG_FRONTEND_GPIO_REG_PTR, (mask_to_set ^ mask_of_pins));
+	nrf_gpio_pin_set(NRF_GPIO_PIN_MAP(CONFIG_LOG_FRONTEND_PORT_NUMBER, pin_nums->CLK));
+}
+
+static inline void log_frontend_0(const char *str, struct log_msg_ids src_level)
+{
+	unsigned int key = irq_lock();
+
+	nrf_gpio_port_dir_output_set(LOG_FRONTEND_GPIO_REG_PTR, mask_of_pins_log);
```

Comment:
this should be done only once. Probably there should be something like log_frontend_init() in the API.